src/algorithms/calorimetry/TrackClusterMergeSplitter.cc

```diff
@@ -0,0 +1,460 @@
+// Copyright (C) 2024 Derek Anderson
+
+#define EICRECON_TRACKCLUSTERMERGESPLITTER_CC
+
+#include <cmath>
+#include <regex>
+#include <utility>
+#include <algorithm>
+// dd4hep utilities
+#include <DD4hep/Readout.h>
+// jana utilities
+#include <JANA/JException.h>
+// edm4hep types
+#include <edm4hep/Vector3f.h>
+#include <edm4hep/utils/vector_utils.h>
+// edm4eic types
+#include <edm4eic/ProtoCluster.h>
+#include <edm4eic/TrackSegment.h>
+#include <edm4eic/CalorimeterHit.h>
+
+// algorithm definition
+#include "TrackClusterMergeSplitter.h"
+
+
+
+namespace eicrecon {
+
+  // --------------------------------------------------------------------------
+  //! Initialize algorithm
+  // --------------------------------------------------------------------------
+  void TrackClusterMergeSplitter::init(
+    const dd4hep::Detector* detector,
+    const dd4hep::rec::CellIDPositionConverter* converter
+  ) {
+
+    m_detector = detector;
+    m_converter = converter;
+
+  }  // end 'init(dd4hep::Detector*)'
+
+
+
+  // --------------------------------------------------------------------------
+  //! Process inputs
+  // --------------------------------------------------------------------------
+  void TrackClusterMergeSplitter::process(
+    const TrackClusterMergeSplitter::Input& input,
+    const TrackClusterMergeSplitter::Output& output
+  ) const {
+
+    // grab inputs/outputs
+    const auto [in_protoclusters, in_projections] = input;
+    auto [out_protoclusters] = output;
+
+    // exit if no clusters in collection
+    if (in_protoclusters -> size() == 0) {
+      throw JException("No proto-clusters in input collection!");
+    }
+
+    // reset bookkeeping containers
+    reset_bookkeepers();
+
+    // flag all clusters as not being consumed
+    for (auto cluster : *in_protoclusters) {
+      m_mapIsConsumed.insert(
+        {cluster.getObjectID().index, false}
+      );
+    }
+    trace("Initialized map of consumed clusters");
+
+    // collect relevant projections
+    get_projections(in_projections, (*in_protoclusters)[0].getHits(0));
+
+    // match clusters to projections
+    if (m_vecProject.size() == 0) {
+      debug("No projections to match clusters to.");
+    } else {
+      match_clusters_to_tracks(in_protoclusters);
+    }
+
+    // determine what clusters to merge
+    for (auto clustAndProject : m_mapClustProject) {
+
+      // skip if cluster is already used
+      if (m_mapIsConsumed[clustAndProject.first]) {
+        continue;
+      }
+
+      // grab matched cluster-projections
+      auto clustSeed = (*in_protoclusters)[clustAndProject.first];
+      auto projSeed = m_vecProject[clustAndProject.second];
+
+      // add cluster to lists and flag as used
+      m_mapClustToMerge.insert(
+        {clustAndProject.first, {}}
+      );
+      m_mapProjToMerge.insert(
+        {clustAndProject.first, {clustAndProject.second}}
+      );
+      m_mapIsConsumed[clustAndProject.first] = true;
```

Comment:
Just use `std::set`/`std::unordered_set`.

Reply:
I'm a little unclear on how I might use `std::set` here... Would this be for `m_mapIsConsumed` or for `m_mapProjToMerge`? (Or both?)

Reply:
Just for `m_mapIsConsumed`. There is no point to store "true", the mere presence in a set carries that information.

Reply:
Ah! I see! Gotcha!